nbgrader/converters/base.py

```diff
@@ -359,3 +363,28 @@ def _handle_failure(gd):
 
             self.log.error(msg)
             raise NbGraderException(msg)
+
+    def recursively_archive(self, directory, archive_name):
+        all_files_and_folders = [ item for item in os.listdir(directory) if item != archive_name+".zip" ]
+
+        for item in all_files_and_folders:
+            full_path = os.path.join(directory, item)
+            if os.path.isfile(full_path):
+                self.archive.write(os.path.join(self.writer.build_directory, full_path.split(archive_name, 1)[-1][1:]), archive_name+"/"+full_path.split(archive_name, 1)[-1][1:])
+            elif os.path.isdir(full_path):
+                self.recursively_archive(full_path, archive_name)
+
+    def make_archive(self):
+        current_dir = self.writer.build_directory
+        archive_name = os.path.basename(self.writer.build_directory)
+
+        if archive_name+".zip" in os.listdir(current_dir):
+            os.remove(current_dir+"/"+archive_name+".zip")
+            self.log.info("Zip file already exists. File removed.")
+
+        self.archive = zipfile.ZipFile(current_dir+"/"+archive_name+".zip", 'w', compression=zipfile.ZIP_DEFLATED)
```

Comment:
The only thing that should have access to the archive is `recursively_archive`, so I would pass the handle to the archive into that function as an argument rather than setting it as an attribute of the class. You should also probably wrap the archiving funcationality with a try/except/finally so you can still close the archive even if there is an error.